Dashboard: Fix Annotation runtime error when a data source does not support annotations

axelavargas opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this PR about?**

A runtime error occurred when a default data source with no annotation support was configured. In this pull request, I am handling the runtime error and informing the user the current data source is not supported.


#### Before:
`TypeError: this.angularComponent.getScope(...).$watch is not a function`

![image](https://github.com/user-attachments/assets/77701f4d-17dc-4ebe-8444-ea8b026e1fe7)


#### After:

![AnnotationDSNotSupported](https://github.com/user-attachments/assets/cb898521-3184-42ad-83b5-4c5a96914368)


**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Related [#11948](https://github.com/grafana/support-escalations/issues/11948)

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

axelavargas committed:
Return grafana ds when default annotation datasource does not support annotations

axelavargas committed:
Add extra test case

axelavargas committed:
Merge remote-tracking branch 'origin' into axelav/annotation-query-runtime-error

ivanortegaalba commented on public/app/features/dashboard-scene/settings/AnnotationsEditView.tsx:
```diff
@@ -51,11 +51,32 @@ export class AnnotationsEditView extends SceneObjectBase<AnnotationsEditViewStat
+    // get current default datasource ref from instance settings
+    const defaultInstanceDS = getDataSourceSrv().getInstanceSettings(null);
+    // get datasource ref from default datasource
+    const defaultDSRef = getDataSourceRef(defaultInstanceDS!);
```
I would not assume there is always a default instance here. Either the TS is wrong, or there is a possibility in code to get a falsy value here. So, following the implementation, I'd control what happens in case this is a falsy value. WDYT?

ivanortegaalba reviewed (approved):
Looks good to me Alexa! Just an small comment about assuming there is always a default DS. Otherwise LGTM!

torkelo commented on public/app/features/dashboard-scene/settings/AnnotationsEditView.tsx:
```diff
@@ -51,11 +51,32 @@ export class AnnotationsEditView extends SceneObjectBase<AnnotationsEditViewStat
+    const dataSourceSrv = await getDataSourceSrv().get(defaultDSRef);
+
+    // if default ds does not support annotations, return the grafana datasource
+    if (!dataSourceSrv?.annotations) {
+      const dsGrafana = getDataSourceSrv().getInstanceSettings('-- Grafana --');
+      if (!dsGrafana) {
+        throw new Error('Grafana datasource not found');
+      }
+      const grafanaDataSourceRef = getDataSourceRef(dsGrafana);
+      return grafanaDataSourceRef;
+    } else {
+      return defaultDSRef;
+    }
```
there is an annotations flag in plugin json if the datasource supports annotations so you should be able to check that. 

that flag should be on the defaultInstanceDS (InstanceSettings) 

that is better than checking dataSourceSrv.annotations as there are older data sources that implement it via the old annotationQuery function that is directly on the DataSource class

torkelo commented:
question, does this error happen in the old arch as well? (noticed that there we just use the default data source as well), no that that we need to fix it there as well , just curious why the old annotations "create new query" logic managed to live without this complexity

axelavargas commented:
> question, does this error happen in the old arch as well? (noticed that there we just use the default data source as well), no that that we need to fix it there as well , just curious why the old annotations "create new query" logic managed to live without this complexity

In the old arch, it fails gracefully. For instance, for my default datasource "test-zipkin" , In the terminal I saw this:

`'Failed to find exported plugin component for annotations-query-ctrl-zipkin`

That resulted in no Query Editor in the UI.

axelavargas committed:
Refactor: Instead of default DS fail gracefully and show error message to users

axelavargas committed:
Implement same behaviour in old arch

axelavargas committed:
Add spyOn on console.error

axelavargas commented:
@torkelo @ivanortegaalba , I have changed the approach to solving this problem, instead of doing a fallback to the `Grafana` data source, I am showing an inline error to the user.  It fails gracefully (similar to the old architecture).

I also added this error in the old architecture.

axelavargas commented on public/app/features/dashboard-scene/settings/AnnotationsEditView.tsx:
```diff
@@ -51,11 +51,23 @@ export class AnnotationsEditView extends SceneObjectBase<AnnotationsEditViewStat
   }
 
-  public onNew = () => {
+  public getDataSourceRefForAnnotation = async () => {
```
oh yes! left over from the old approach, good catch!

torkelo commented on public/app/features/dashboard-scene/settings/annotations/AnnotationSettingsEdit.tsx:
```diff
@@ -183,6 +184,14 @@ export const AnnotationSettingsEdit = ({ annotation, editIndex, panels, onUpdate
         <Field label="Data source" htmlFor="data-source-picker">
           <DataSourcePicker annotations variables current={annotation.datasource} onChange={onDataSourceChange} />
         </Field>
+        {!ds ||
+          (!ds.annotations && (
```
should this not check meta.annotations as well? 

{ds?.meta.annotations && (

axelavargas replied:
indeed, I keep forgetting what you mentioned regarding old data sources 🙈

axelavargas committed:
Clean up and use ds.meta.annotation for error

axelavargas committed:
Update translation
